src/msg/consumer/types.go

```diff
@@ -132,17 +132,58 @@ type MessageProcessor interface {
 	Close()
 }
 
-// NewMessageProcessorFn creates a new MessageProcessor scoped to a single connection. Messages are processed serially
-// in a connection.
-type NewMessageProcessorFn func() MessageProcessor
+// MessageProcessorPool returns MessageProcessors.
+type MessageProcessorPool interface {
```

Comment:
IMO a bit confusing using the name "pool" here since it isn't actually pooling anything, just controlling if the processor provided process itself should get closed or not.

Reply:
well the singleton is definitely pooling. do you have another naming suggestion?

Reply:
Maybe change to `Provider` - also maybe remove `Put` and just have the provider differences be: singleton wraps a processor such that it's `Close` is no-op; otherwise provider gives a new processor that has a real `Close`.

Reply:
yea i originally had a wrapper that did nothing on Close. To me that seemed weird since it was so coupled the `Handler` impl. But it seems like this pool thingy is confusing, so i'll just go back to that previous impl.